universal-application-tool-0.0.1/app/services/cloud/azure/BlobStorage.java

```diff
@@ -0,0 +1,241 @@
+
+      BlobSasPermission blobSasPermission =
+          new BlobSasPermission().setReadPermission(true).setWritePermission(true);
+
+      BlobServiceSasSignatureValues signatureValues =
+          new BlobServiceSasSignatureValues(
+                  OffsetDateTime.now(zoneId).plus(AZURE_SAS_TOKEN_DURATION), blobSasPermission)
+              .setProtocol(SasProtocol.HTTPS_ONLY);
+
+      return blobClient.generateUserDelegationSas(signatureValues, getUserDelegationKey());
+    }
+
+    @Override
+    public String getBlobUrl(String fileName) {
+      BlobClient blobClient =
+          blobServiceClient.getBlobContainerClient(container).getBlobClient(fileName);
+      return blobClient.getBlobUrl();
+    }
+  }
+
+  class AzuriteClient implements Client {
+    // Using Account SAS because Azurite emulator does not support User Delegation SAS yet
+    // See https://github.com/Azure/Azurite/issues/656
+
+    private static final String AZURE_LOCAL_CONNECTION_STRING_CONF_PATH = "azure.local.connection";
+
+    private final String connectionString;
+    private final BlobServiceClient blobServiceClient;
+    private final BlobContainerClient blobContainerClient;
+
+    AzuriteClient(Config config) {
+      this.connectionString =
+          checkNotNull(config).getString(AZURE_LOCAL_CONNECTION_STRING_CONF_PATH);
+      this.blobServiceClient =
+          new BlobServiceClientBuilder().connectionString(connectionString).buildClient();
+
+      setCorsRules(blobServiceClient);
+      this.blobContainerClient = blobServiceClient.getBlobContainerClient(container);
+      if (!blobContainerClient.exists()) {
+        blobContainerClient.create();
+      }
+    }
+
+    private void setCorsRules(BlobServiceClient blobServiceClient) {
+      BlobServiceProperties properties =
+          new BlobServiceProperties()
+              .setCors(
+                  List.of(
+                      new BlobCorsRule()
+                          // TODO: Make these more specific
+                          .setAllowedOrigins("*")
+                          .setAllowedHeaders("*")
+                          .setExposedHeaders("*")
+                          .setAllowedMethods("GET,PUT,OPTIONS")
+                          .setMaxAgeInSeconds(500)));
+      blobServiceClient.setProperties(properties);
+    }
+
+    @Override
+    public String getSasToken(String fileName) {
+      BlobClient blobClient = blobContainerClient.getBlobClient(fileName);
+
+      BlobSasPermission blobSasPermission =
+          new BlobSasPermission()
+              .setReadPermission(true)
+              .setWritePermission(true)
+              .setListPermission(true);
+
+      BlobServiceSasSignatureValues signatureValues =
+          new BlobServiceSasSignatureValues(
+                  OffsetDateTime.now(zoneId).plus(AZURE_SAS_TOKEN_DURATION), blobSasPermission)
+              .setProtocol(SasProtocol.HTTPS_HTTP); // TODO: Get this to work with HTTPS_ONLY
```

Comment:
thought (non-blocking): we can't use HTTPS in dev so this may not be a possible TODO